src/axom/core/Array.hpp

```diff
@@ -14,12 +14,57 @@
 #include "axom/core/IteratorBase.hpp"         // for Iterator
 
 // C/C++ includes
+#include <array>     // for std::array
 #include <iostream>  // for std::cerr and std::ostream
+#include <numeric>   // for std::inner_product
 
 namespace axom
 {
-// Forward declare the templated classes and operator function(s)
+// TODO: Add this as a non-type template parameter to Array/View
+// The intent is that there will also be a "Dynamic" or "Polymorphic"
+// resource type
+// enum MemoryResourceType
+// {
+//   Host,
+//   Device,
+//   Unified,
+//   Pinned,
+//   Constant,
+//   File,
+//   NoOp,
+//   Shared,
+//   Unknown
+// };
+
+namespace detail
+{
+// Takes the product of a parameter pack of IndexTypes
 template <typename T>
+IndexType packProduct(T&& t)
+{
+  return t;
+}
+
+template <typename T, typename... Args>
+IndexType packProduct(T&& t, Args... args)
+{
+  return t * packProduct(args...);
+}
+
+template <typename... Args>
+bool allPositive(Args... args)
+{
+  // Use the "expander trick" because we don't have fold expressions
+  using expander = int[];
+  bool result = true;
+  (void)expander {(result = result && (args > 0), 0)...};
+  return result;
+}
```

Comment:
these are a little bit hard to understand, does this construction work?

```cpp
template < int n >
int product_of(const int (&arr)[n]) {
  int p = 1;
  for (int i = 0; i < n; i++) { p *= arr[i]; }
  return p;
}

template < int n >
bool all_positive(const int (&arr)[n]) {
  for (int i = 0; i < n; i++) { if (arr[i] <= 0) return false; }
  return true;
}

template < typename ... T >
int foo(T ... args) {
  assert(all_positive({int(args) ...}));
  return product_of({int(args) ... });
}
```

https://godbolt.org/z/1569T57nK

Reply:
Thanks for the suggestion - it looks like despite a little more setup at the call site it's less instructions overall too: https://godbolt.org/z/9PsoTzGjE